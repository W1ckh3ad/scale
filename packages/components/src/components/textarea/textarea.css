.input {
  position: relative;
}
.input .input__helper-text,
.input .input__counter {
  font-weight: var(--scl-font-weight-bold);
}

.input .input__textarea {
  width: 100%;
  margin: 0;
  resize: vertical;
  display: flex;
  outline: none;
  padding: var(--scl-spacing-12) var(--scl-spacing-12);
  padding-top: var(--scl-spacing-24);
  z-index: 1;
  box-sizing: border-box;
  transition: all var(--scl-motion-duration-fast)
    var(--scl-motion-easing-standard);
  font-family: inherit;
  font-size: var(--scl-font-size-16);
  border-radius: var(--scl-radius-4);
  border: var(--scl-spacing-1) solid var(--scl-color-text-standard);
}
.input .input__counter {
  display: flex;
  transition: all var(--scl-motion-duration-fast)
    var(--scl-motion-easing-standard);
  margin-left: auto;
  padding-right: var(--scl-spacing-12);
  justify-content: flex-end;
  font-size: var(--scl-font-size-12);
  line-height: var(--scl-font-line-height-133);
  color: var(--scl-color-text-standard);
}
.input .input__helper-text {
  transition: all var(--scl-motion-duration-fast)
    var(--scl-motion-easing-standard);
  padding-left: var(--scl-spacing-12);
  font-size: var(--scl-font-size-12);
  line-height: var(--scl-font-line-height-133);
  color: var(--scl-color-blue-70);
}
.input .input__meta {
  display: flex;
  justify-content: space-between;
  margin-top: var(--scl-spacing-4);
}
.input:not(.input--disabled) .input__textarea:hover,
.input:not(.input--disabled) .input__textarea:focus {
  border-color: var(--scl-color-primary-hover, #f90984);
}
.input:not(.input--disabled) .input__textarea:focus {
  box-shadow: 0 0 0 var(--scl-spacing-2) var(--scl-color-focus);
}
.input:not(.input--disabled) .input__textarea:focus::placeholder {
  color: var(--scl-color-grey-60);
  transition: all var(--scl-motion-duration-fast)
    var(--scl-motion-easing-standard);
}
.input .input__textarea::placeholder,
.input ::placeholder {
  color: transparent;
  transition: all var(--scl-motion-duration-fast)
    var(--scl-motion-easing-standard);
}

.input__label {
  top: 0;
  left: 0;
  color: var(--scl-color-grey-60);
  display: flex;
  z-index: var(--scl-z-index-10);
  position: absolute;
  transition: all var(--scl-motion-duration-fast)
    var(--scl-motion-easing-standard);
  pointer-events: none;
  font-size: var(--scl-font-size-16);
  /* previously: transform: translate(13px, calc((48px - 1rem) / 2)); */
  transform: translate(
    var(--scl-spacing-12),
    calc((var(--scl-spacing-48) - var(--scl-font-size-16)) / 2)
  );
  font-weight: var(--scl-font-weight-medium);
}
.input--has-focus .input__label,
.animated .input__label {
  color: var(--scl-color-grey-60);
  transform: translate(var(--scl-spacing-12), var(--scl-spacing-8));
  transition: all var(--scl-motion-duration-fast)
    var(--scl-motion-easing-standard);
  font-size: var(--scl-font-size-10);
  font-weight: var(--scl-font-weight-bold);
}
.input--status-error .input__textarea {
  border: var(--scl-spacing-2) solid var(--scl-color-background-error);
}
.input--status-error .input__helper-text {
  color: var(--scl-color-text-error);
}
.input--status-error .input__counter {
  color: var(--scl-color-text-error);
}
.input--size-small .input__label {
  top: 0;
  left: 0;
  color: var(--scl-color-grey-60);
  display: flex;
  z-index: var(--scl-z-index-10);
  position: absolute;
  transition: all var(--scl-motion-duration-fast)
    var(--scl-motion-easing-standard);
  pointer-events: none;
  font-size: var(--scl-font-size-16);
  /* previously: transform: translate(
    13px,
    calc((var(--input-small-height, 40px) - 1rem) / 2)
  ); */
  transform: translate(
    var(--scl-spacing-12),
    calc((var(--scl-spacing-40) - var(--scl-font-size-16)) / 2)
  );
  font-weight: var(--scl-font-weight-medium);
}
.input--size-small.input--has-focus .input__label,
.input--size-small.animated .input__label {
  color: var(--scl-color-grey-60);
  transform: translate(var(--scl-spacing-12), var(--scl-spacing-4));
  transition: all var(--scl-motion-duration-fast)
    var(--scl-motion-easing-standard);
  font-size: var(--scl-font-size-10);
  font-weight: var(--scl-font-weight-medium);
}
.input--transparent .input__textarea {
  background-color: transparent;
}
.input__textarea-label-safety-background {
  top: var(--scl-spacing-2);
  left: var(--scl-spacing-2);
  right: var(--scl-spacing-2);
  position: absolute;
  pointer-events: none;
  border-radius: var(--scl-radius-4);
  height: var(--scl-spacing-24);
  background-color: var(--scl-color-background-standard, #ffffff);
}
.input--transparent .input__textarea-label-safety-background,
.input--disabled .input__textarea-label-safety-background {
  background-color: transparent;
}
.input--disabled label,
.input--disabled .input__label,
.input--disabled input,
.input--disabled .input__textarea,
.input--disabled .input__helper-text {
  cursor: not-allowed;
  border-color: var(--scl-color-background-disabled);
  color: var(--scl-color-background-disabled);
  background: var(--scl-color-white);
}
