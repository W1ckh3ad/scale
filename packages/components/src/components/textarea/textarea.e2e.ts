import { newE2EPage } from '@stencil/core/testing';

describe('scale-textarea', () => {
  it('renders', async () => {
    const page = await newE2EPage();
    await page.setContent('<scale-textarea/>');
    const element = await page.find('scale-textarea');
    expect(element).toHaveClass('hydrated');
  });
});
