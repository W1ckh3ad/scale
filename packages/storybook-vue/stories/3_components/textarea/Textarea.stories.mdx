import { Meta, Story, Props, Preview, Description } from "@storybook/addon-docs/blocks";
import ScaleTextarea from "./ScaleTextarea.vue";
import { action } from "@storybook/addon-actions";
export const ICON = "M16.471,8.321 L12.613594,0.502886112 L8.755,8.321 L0.126435258,9.57533801 L6.37,15.66 L4.89610548,24.2548735 L12.613,20.197 L20.3310826,24.2548735 L18.856,15.66 L25.1007528,9.57533801 L16.471,8.321 Z M9.41891527,9.23556376 L12.613,2.761 L15.8082728,9.23556376 L22.951,10.273 L17.7826928,15.3122039 L19.003,22.427 L12.613594,19.067774 L6.223,22.427 L7.44449521,15.3122039 L2.275,10.273 L9.41891527,9.23556376 Z";

<Meta
  title="Components/Textarea"
  component={ScaleTextarea}
  argTypes={{
    resize: {
      control: {
        type: "select",
        options: ["unset", "none", "vertical", "horizontal"]
      }
    },
    type: {
      control: {
        type: "select",
        options: [
          "email",
          "hidden",
          "number",
          "password",
          "tel",
          "text",
          "checkbox",
          "radio",
          "select",
          "textarea",
          "url"
        ]
      }
    },
    variant: {
      control: { type: "select", options: ["animated", "static"] }
    }
  }}
/>

export const Template = (args, { argTypes }) => ({
  components: { ScaleTextarea },
  props: {
    label: String,
    ...ScaleTextarea.props
  },
  template: `
    <scale-input
      :name="name"
      :variant="variant"
      :type="type"
      :label="label"
      :size="size"
      :rows="rows"
      :cols="cols"
      :helper-text="helperText"
      :status="status"
      :max-length="maxLength"
      :min-length="minLength"
      :placeholder="placeholder"
      :disabled="disabled"
      :required="required"
      :counter="counter"
      :resize="resize"
      :value="value"
      :input-id="inputId"
      :transparent="transparent"
      @scaleChange="scaleChange"
      @scaleFocus="scaleFocus"
      @scaleBlur="scaleBlur"
      @scaleKeyDown="scaleKeyDown"
    ></scale-input>
  `,
  methods: {
    scaleChange: action("scaleChange"),
    scaleFocus: action("scaleFocus"),
    scaleBlur: action("scaleBlur"),
    scaleKeyDown: action("scaleKeyDown"),
  }
})

# Textarea

## Standard

<Preview withSource="none">
  <Story
    name="Standard"
    args={{
      label: "Textarea",
      type: "textarea",
      rows: 3,
      resize: "vertical",
      variant: 'animated'
    }}
  >
    {Template.bind({})}
  </Story>
</Preview>

<Props story="Standard" />

```html
<scale-input variant="animated" label="Textarea" type="textarea" rows="3" resize="vertical"></scale-input>
```



## Placeholder

<Preview withSource="none">
  <Story
    name="Placeholder"
    args={{
      type: "textarea",
      label: "Placeholder",
      placeholder: "this is the placeholder",
      variant: 'animated'
    }}
  >
    {Template.bind({})}
  </Story>
</Preview>

```html
<scale-input variant="animated" label="Placeholder" type="textarea" :placeholder="this is the placeholder"></scale-input>
```


## Helper Text

<Preview withSource="none">
  <Story
    name="Helper Text"
    args={{
      type: "textarea",
      helperText: "Make sure to fill this",
      label: "Helper Text",
      variant: 'animated'
    }}
  >
    {Template.bind({})}
  </Story>
</Preview>

```html
<scale-input variant="animated" type="textarea" label="Helper Text" helperText="Make sure to fill this"></scale-input>
```

## With Error

<Preview withSource="none">
  <Story
    name="With Error"
    args={{
      type: "textarea",
      label: "With Error",
      status: "error",
      helperText: "something is wrong",
      variant: 'animated'
    }}
  >
    {Template.bind({})}
  </Story>
</Preview>

```html
<scale-input variant="animated" type="textarea" label="With Error" helper-text="something is wrong" status="error"></scale-input>
```


## Disabled

<Preview withSource="none">
  <Story
    name="Disabled"
    args={{
      type: "textarea",
      label: "Disabled",
      disabled: true,
      variant: 'animated'
    }}
  >
    {Template.bind({})}
  </Story>
</Preview>

```html
<scale-input variant="animated" type="textarea" label="Disabled" disabled></scale-input>
```


## Max Length With Counter

<Preview withSource="none">
  <Story
    name="Max Length With Counter"
    args={{
      label: "Max Length With Counter",
      maxLength: 10,
      counter: true,
      type: "textarea",
      variant: 'animated'
    }}
  >
    {Template.bind({})}
  </Story>
</Preview>

```html
<scale-input variant="animated" type="textarea" label="Max Length With Counter" max-length="10" counter></scale-input>
```


## More rows

<Preview withSource="none">
  <Story
    name="More rows"
    args={{
      label: "More rows",
      type: "textarea",
      rows: 10,
      variant: 'animated'
    }}
  >
    {Template.bind({})}
  </Story>
</Preview>

```html
<scale-input variant="animated" type="textarea" label="More rows" rows="10"></scale-input>
```
