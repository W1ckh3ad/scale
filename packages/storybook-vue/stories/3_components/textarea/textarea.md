# Textarea

With a text area, users can enter long text in a form.

## General

### When to Use

Use text area whenever users should be able to enter a larger amount of text. Examples: comments or notes on an order, a message in a contact form.

### When Not to Use

For simple text input that does not require multiple lines, use an input filed instead.

## Elements

![Image Name](/assets/3_components/text-area/image-20200811104309878.png)

### Label (1)

Clearly label your text area so users know which information to enter. e.g. "notes".

### Placeholder Text (2) (Optional)

An example text that is displayed on focus when no input has been entered yet, e.g. "your message".

### Info (3)

Optional instructions about the field, e.g. "this will not be shared with your contacts".

### Character Count (4) (Optional)

Limit the amount of characters for a text area. When a limit is set, a counter is displayed e.g. "14/75".

### Error / Improve / Required Input (5) (Optional)

Error text can be set to highlight the changes a user has to make for the information to validate. e.g. "please enter a valid e-mail adress". The error text overwrites the information notes, if present.

For required fields, use this flag to display the corresponding information and validate the form accordingly.

### Resize (6) (Optional)

With this element, the user can adjust the size of the text field to make the entire text visible.

## Related Components

Input Field
