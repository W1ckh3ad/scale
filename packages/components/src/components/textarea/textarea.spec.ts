/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import { newSpecPage } from '@stencil/core/testing';
import { Textarea } from './textarea';
describe('Textarea', () => {
  it('should match snapshot', async () => {
    const page = await newSpecPage({
      components: [Textarea],
      html: `<scale-textarea></scale-textarea>`,
    });
    expect(page.root).toMatchSnapshot();
  });
});
