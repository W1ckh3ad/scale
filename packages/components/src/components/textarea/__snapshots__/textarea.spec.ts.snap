// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textarea should match snapshot -> counter 1`] = `
<scale-textarea counter="true">
  <div class="textarea">
    <label class="textarea__label" htmlfor="input-textarea2"></label><textarea aria-describedby="helper-message-3" class="textarea__control" id="input-textarea2" value=""></textarea>
    <span aria-hidden="true" class="textarea__label-safety-background"></span>
    <div aria-live="polite" aria-relevant="additions removals" class="textarea__meta" id="helper-message-3">
      <div class="textarea__counter">
        0 /
      </div>
    </div>
  </div>
</scale-textarea>
`;

exports[`Textarea should match snapshot -> helpertext 1`] = `
<scale-textarea helpertext="helpertext">
  <div class="textarea">
    <label class="textarea__label" htmlfor="input-textarea1"></label><textarea aria-describedby="helper-message-2" class="textarea__control" id="input-textarea1" value=""></textarea>
    <span aria-hidden="true" class="textarea__label-safety-background"></span>
  </div>
</scale-textarea>
`;

exports[`Textarea should match snapshot 1`] = `
<scale-textarea>
  <div class="textarea">
    <label class="textarea__label" htmlfor="input-textarea0"></label><textarea aria-describedby="helper-message-1" class="textarea__control" id="input-textarea0" value=""></textarea>
    <span aria-hidden="true" class="textarea__label-safety-background"></span>
  </div>
</scale-textarea>
`;
